Improve usage of `String.split()` vs `get_slice()`

KoBeWi opened the pull request:
I noticed that we have this in some places: `string.split("/")[0]`. It creates a temporary array and takes first element. It's extremely wasteful and we have a better method for that: `string.get_slice("/", 0)`. So I wanted to replace the single-splits, but obviously I found more cases where this can be improved and the PR ended up a bit big...

From my testing, doing `get_slice_count()` + `get_slice()` * 2 is still faster than splitting, so I did a couple of changes based on that, but there is more code that can be improved. IMO we should avoid splitting, unless it's used for iterating. The performance gain is marginal though, I guess.

Other than this, I changed `get_slice()` to return empty string for slice index > 0 if there is no delimiter in the string. This is more intuitive behavior. Also improved docs for `get_slice()` methods, as they were giving a wrong impression that these methods split the String, which is not true.

I can do a follow-up PR that replaces more splits.
And I'm planning to do an opposite PR that removes `get_slice()` iterations, because it's bad xd `get_slice()` is only if you need to access one or 2 strings, not loop through them.

akien-mga commented on core/string/ustring.cpp:
```diff
@@ -1032,7 +1032,11 @@ String String::get_slice(String p_splitter, int p_slice) const {
 	}
 	if (find(p_splitter) == -1) {
-		return *this;
+		if (p_slice > 0) {
```
Nitpick but you could remove the commented out `slices` above.

akien-mga commented on doc/classes/String.xml:
```diff
@@ -222,28 +222,35 @@
 			<description>
-				Splits a string using a [code]delimiter[/code] and returns a substring at index [code]slice[/code]. Returns an empty string if the index doesn't exist.
-				This is a more performant alternative to [method split] for cases when you need only one element from the array at a fixed index.
+				Splits a string using a [code]delimiter[/code] and returns a substring at index [code]slice[/code]. Returns the whole string if delimiter doesn't exist in the string and [code]slice[/code] is [code]0[/code]. Returns an empty string if the slice index doesn't exist.
```
```suggestion
				Splits a string using a [code]delimiter[/code] and returns a substring at index [code]slice[/code]. Returns the whole string if the delimiter doesn't exist in the string and [code]slice[/code] is [code]0[/code]. Returns an empty string if the slice index doesn't exist.
```

akien-mga commented on editor/editor_inspector.cpp:
```diff
@@ -2464,27 +2464,17 @@ void EditorInspector::update_tree() {
-			} else {
-				section_depth = 0;
-			}
+			subgroup_base = p.hint_string.get_slice(",", 0);
```
Wrong change here:
```suggestion
			group_base = p.hint_string.get_slice(",", 0);
```

KoBeWi replied:
Copy-pasted too hard .-.

akien-mga commented on editor/plugins/tiles/tile_set_atlas_source_editor.cpp:
```diff
@@ -789,9 +789,9 @@ void TileSetAtlasSourceEditor::_update_current_tile_data_editor() {
-		if (components.size() >= 1) {
-			property = components[0];
+		const String &selected = tile_inspector->get_selected_path();
+		if (selected.get_slice_count("/") >= 1) {
```
I think `get_slice_count()` always returns 1 or more, no? So the check is unnecessary.

akien-mga commented on modules/visual_script/editor/visual_script_editor.cpp:
```diff
@@ -912,15 +911,9 @@ void VisualScriptEditor::_update_graph(int p_only_id) {
-							} else {
-								current_val = 1 << i;
-							}
+							uint32_t current_val = option.get_slice(":", 1).to_int();
```
What if there's only one slice?

KoBeWi replied:
Then it's 0.
The previous code didn't handle it really well either; checking size for -1 makes no sense.

akien-mga commented on modules/visual_script/editor/visual_script_editor.cpp:
```diff
@@ -3487,106 +3480,105 @@ void VisualScriptEditor::_selected_connect_node(const String &p_text, const Stri
 				n->set_basic_type(Variant::PACKED_COLOR_ARRAY);
 			}
-			n->set_basic_type_constant(property_path[1]);
+			n->set_basic_type_constant(p_text.get_slice(":", 1));
```
I would extract this together with the `base_type` before the huge list of `if`/`else`, so the code is more self-explanatory:
```
const String base_type = p_text.get_slice(":", 0);
const String class_constant = p_text.get_slice(":", 1);
...
```

akien-mga commented on modules/gdscript/language_server/gdscript_workspace.cpp:
```diff
@@ -615,9 +615,8 @@ const lsp::DocumentSymbol *GDScriptWorkspace::resolve_symbol(const lsp::TextDocu
-		Vector<String> identifier_parts = symbol_identifier.split("(");
-		if (identifier_parts.size()) {
-			symbol_identifier = identifier_parts[0];
+		if (symbol_identifier.get_slice_count("(") > 0) {
```
Apparently not, my bad.

akien-mga commented on modules/visual_script/editor/visual_script_property_selector.cpp:
```diff
@@ -1122,9 +1122,8 @@ TreeItem *VisualScriptPropertySelector::SearchRunner::_create_class_item(TreeIte
-		Vector<String> path = p_doc->name.split("/");
 		icon = ui_service->get_theme_icon(SNAME("VisualScript"), SNAME("EditorIcons"));
-		text_0 = path[path.size() - 1];
+		text_0 = p_doc->name.get_slice("/", p_doc->name.get_slice_count("/") - 1);
```
I'm not sure that's equivalent because `get_slice_count("/")` can return 0 if `p_doc->name` is empty (btw this could reuse `details` defined above).

So this could end up with `-1`, while `split("/")` would always have 1 item.

So... this whole PR needs to be re-reviewed in light of this difference.

KoBeWi replied:
Well, then `text_0` will be an empty String (`get_slice()` returns "" for -1), so it's not a problem.

KoBeWi commented on editor/editor_inspector.cpp:
```diff
@@ -2464,27 +2464,17 @@ void EditorInspector::update_tree() {
-				section_depth = 0;
-			}
+			subgroup_base = p.hint_string.get_slice(",", 0);
+			section_depth = p.hint_string.get_slice(",", 1).to_int();
```
> is it relying on get_slice(",", 1) to return "" and converting that to int would become 0?

Yes. I will change it back :/

KoBeWi committed:
Improve usage of String.split() vs get_slice()

KoBeWi commented:
Rebased. I removed the `get_slice()` behavior change and all code relying on it. I also discarded all conflicting changes and fixed some newly added trivial cases.

akien-mga reviewed (approved):
Seems fine to me. Will need a rebase, maybe shortly after the 4.3 release to merge for 4.4.

Calinou reviewed (approved):
Code looks good to me. Should be good to merge after rebasing (rebase is only needed after 4.3 is released).

groud commented on editor/plugins/tiles/tile_set_atlas_source_editor.cpp:
```diff
@@ -872,9 +872,9 @@ void TileSetAtlasSourceEditor::_update_current_tile_data_editor() {
-		Vector<String> components = tile_inspector->get_selected_path().split("/");
-		if (components.size() >= 1) {
-			property = components[0];
+		const String &selected = tile_inspector->get_selected_path();
```
I think I would avoid changing it here for consistency. The file uses this pattern everywhere:
```
Vector<String> components = String(p_property).split("/", true, 2);
```
Not that I mind much, but I think consistency kind of helps maintenance and refactoring.

KoBeWi replied:
I don't really care about consistency if it means creating a temporary array just to fetch 1 element. Also I don't see how this helps.

groud replied:
I mean, I do... Code is a lot simpler to follow when it reuse similar patterns in similar cases.

To be clear, I don't mind much the change to be honest. It just feel like this is basically some unneeded optimization (at least in that place I think it brings no significant performance improvement, in the other cases I don't know). But well, it's not a big deal, we can merge as-is it's ok too.

clayjohn replied:
I agree that consistency is valuable in such big systems. What are the other cases in this file? Perhaps everything can switch to using get_slice()?

KoBeWi replied:
Well, technically my changes are self-consistent. `get_slice()` is used everywhere where no more than 2 sliced values are accessed.

groud replied:
I searched for where I used this pattern (with the components vector), this is the list:
![Copie d'écran_20240813_160941](https://github.com/user-attachments/assets/451db4e2-6d4b-4a1e-847f-ac704f46f308)

To be fair, I think the change is OK as part of such a big refactoring PR. But I can't promise to think about keeping things that way the next time I have to refactor something in the TileMap/Set area. :sweat_smile:

clayjohn replied:
@KoBeWi Do you want to switch these cases over to get_slice? too, then we can have consistency and performance :)

KoBeWi replied:
They all use the `maxsplit` argument of `split()`, which don't exist in `get_slice()`. The inspector one loops over the strings, so slicing is justified.

Looks like the code here is the only component split this way 🤔

KoBeWi replied:
Changed to old code with `max_splits`.

KoBeWi commented:
Rebased and replaced with `get_slicec()`. I only updated the files I edited before, didn't look for new replacements.

AThousandShips commented:
`'c'` is `char`, not `char *`

KoBeWi commented:
Looks like we could have overload for `char`, like other methods.

AThousandShips commented:
I've looked at adding a `get_slice_countc` but didn't find it justified enough, can make a PR for it though for ease and alignment, can open a PR for it when I'm back

Calinou reviewed (approved):
Code looks good to me.

Calinou commented on doc/classes/String.xml:
```diff
@@ -341,6 +341,11 @@
 			<param index="0" name="delimiter" type="String" />
 			<description>
 				Returns the total number of slices when the string is split with the given [param delimiter] (see [method split]).
+				Use [method get_slice] to extract a specific slice.
```
Explain why it's better to use `get_slice()`:

```suggestion
				Use [method get_slice] to extract a specific slice, which is more efficient.
```
